File: sdk/ios/Sources/SenseCraftVoiceIOS/BLE/ClipFileData.swift
import Foundation

public let kClipFrameData: UInt8 = 0x01
public let kClipFrameFileStart: UInt8 = 0x10
public let kClipFrameFileEnd: UInt8 = 0x11
public let kClipFrameTransferDone: UInt8 = 0x12
public let kClipDataHeaderSize = 5

public enum ClipFileDataParsed {
    case raw(Data)
    case data(seq: Int, payload: Data)
    case fileStart(filename: String, fileSize: Int)
    case fileEnd(crc32: UInt32)
    case transferDone(sessionId: String, fileCount: Int)
    case invalid(reason: String)
}

public func parseClipFileDataNotify(_ data: Data) -> ClipFileDataParsed {
    guard !data.isEmpty else { return .invalid(reason: "empty") }
    let bytes = [UInt8](data)

    switch bytes[0] {
    case kClipFrameData:
        guard bytes.count >= kClipDataHeaderSize else {
            return .invalid(reason: "DATA short header len=\(bytes.count)")
        }
        let len = Int(bytes[3]) | (Int(bytes[4]) << 8)
        guard bytes.count == kClipDataHeaderSize + len else {
            return .invalid(reason: "DATA len mismatch total=\(bytes.count) payload=\(len)")
        }
        let seq = Int(bytes[1]) | (Int(bytes[2]) << 8)
        return .data(seq: seq, payload: data.subdata(in: kClipDataHeaderSize..<kClipDataHeaderSize + len))

    case kClipFrameFileStart:
        guard bytes.count >= 3 else {
            return .invalid(reason: "FILE_START short len=\(bytes.count)")
        }
        let fnLen = Int(bytes[1])
        guard bytes.count >= 2 + fnLen + 4 else {
            return .invalid(reason: "FILE_START bad fnLen=\(fnLen) total=\(bytes.count)")
        }
        let nameData = data.subdata(in: 2..<2 + fnLen)
        let filename = String(decoding: nameData, as: UTF8.self)
        let metaOff = 2 + fnLen
        let fileSize =
            Int(bytes[metaOff]) |
            (Int(bytes[metaOff + 1]) << 8) |
            (Int(bytes[metaOff + 2]) << 16) |
            (Int(bytes[metaOff + 3]) << 24)
        return .fileStart(filename: filename, fileSize: fileSize)

    case kClipFrameFileEnd:
        guard bytes.count >= 5 else {
            return .invalid(reason: "FILE_END short len=\(bytes.count)")
        }
        let crc =
            UInt32(bytes[1]) |
            (UInt32(bytes[2]) << 8) |
            (UInt32(bytes[3]) << 16) |
            (UInt32(bytes[4]) << 24)
        return .fileEnd(crc32: crc)

    case kClipFrameTransferDone:
        guard bytes.count >= 3 else {
            return .invalid(reason: "TRANSFER_DONE short len=\(bytes.count)")
        }
        let sidLen = Int(bytes[1])
        guard bytes.count >= 2 + sidLen + 4 else {
            return .invalid(reason: "TRANSFER_DONE bad sidLen=\(sidLen)")
        }
        let sidData = data.subdata(in: 2..<2 + sidLen)
        let sessionId = String(decoding: sidData, as: UTF8.self)
        let off = 2 + sidLen
        let fileCount =
            Int(bytes[off]) |
            (Int(bytes[off + 1]) << 8) |
            (Int(bytes[off + 2]) << 16) |
            (Int(bytes[off + 3]) << 24)
        return .transferDone(sessionId: sessionId, fileCount: fileCount)

    default:
        return .raw(data)
    }
}

File: sdk/android/src/main/java/io/sensecraft/voice/android/ble/ClipFileData.kt
package io.sensecraft.voice.android

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

const val kClipFrameData: Int = 0x01
const val kClipFrameFileStart: Int = 0x10
const val kClipFrameFileEnd: Int = 0x11
const val kClipFrameTransferDone: Int = 0x12
const val kClipDataHeaderSize = 5

sealed class ClipFileDataParsed {
    data class Raw(val bytes: ByteArray) : ClipFileDataParsed()
    data class Data(val seq: Int, val payload: ByteArray) : ClipFileDataParsed()
    data class FileStart(val filename: String, val fileSize: Int) : ClipFileDataParsed()
    data class FileEnd(val crc32: Long) : ClipFileDataParsed()
    data class TransferDone(val sessionId: String, val fileCount: Int) : ClipFileDataParsed()
    data class Invalid(val reason: String) : ClipFileDataParsed()
}

fun parseClipFileDataNotify(data: ByteArray): ClipFileDataParsed {
    if (data.isEmpty()) return ClipFileDataParsed.Invalid("empty")
    when (data[0].toInt() and 0xff) {
        kClipFrameData -> {
            if (data.size < kClipDataHeaderSize) return ClipFileDataParsed.Invalid("DATA short header len=${data.size}")
            val len = (data[3].toInt() and 0xff) or ((data[4].toInt() and 0xff) shl 8)
            if (data.size != kClipDataHeaderSize + len) {
                return ClipFileDataParsed.Invalid("DATA len mismatch total=${data.size} payload=$len")
            }
            val seq = (data[1].toInt() and 0xff) or ((data[2].toInt() and 0xff) shl 8)
            return ClipFileDataParsed.Data(seq, data.copyOfRange(kClipDataHeaderSize, kClipDataHeaderSize + len))
        }
        kClipFrameFileStart -> {
            if (data.size < 3) return ClipFileDataParsed.Invalid("FILE_START short len=${data.size}")
            val fnLen = data[1].toInt() and 0xff
            if (data.size < 2 + fnLen + 4) {
                return ClipFileDataParsed.Invalid("FILE_START bad fnLen=$fnLen total=${data.size}")
            }
            val filename = String(data, 2, fnLen, StandardCharsets.UTF_8)
            val off = 2 + fnLen
            val fileSize = ByteBuffer.wrap(data, off, 4).order(ByteOrder.LITTLE_ENDIAN).int
            return ClipFileDataParsed.FileStart(filename, fileSize)
        }
        kClipFrameFileEnd -> {
            if (data.size < 5) return ClipFileDataParsed.Invalid("FILE_END short len=${data.size}")
            val crc = ByteBuffer.wrap(data, 1, 4).order(ByteOrder.LITTLE_ENDIAN).int.toLong() and 0xffffffffL
            return ClipFileDataParsed.FileEnd(crc)
        }
        kClipFrameTransferDone -> {
            if (data.size < 3) return ClipFileDataParsed.Invalid("TRANSFER_DONE short len=${data.size}")
            val sidLen = data[1].toInt() and 0xff
            if (data.size < 2 + sidLen + 4) {
                return ClipFileDataParsed.Invalid("TRANSFER_DONE bad sidLen=$sidLen")
            }
            val sessionId = String(data, 2, sidLen, StandardCharsets.UTF_8)
            val off = 2 + sidLen
            val fileCount = ByteBuffer.wrap(data, off, 4).order(ByteOrder.LITTLE_ENDIAN).int
            return ClipFileDataParsed.TransferDone(sessionId, fileCount)
        }
        else -> return ClipFileDataParsed.Raw(data)
    }
}

